File: Practica de Programacion/Practica 3 Herencia/Ejercicio 3/3 Persona.py
from datetime import date

class Persona:
    def __init__(self, ci, nombre, apellido, fecha_nac):
        self.ci = ci
        self.nombre = nombre
        self.apellido = apellido
        self.fecha_nac = fecha_nac
    
    def edad(self):
        hoy = date.today()
        return hoy.year - self.fecha_nac.year - ((hoy.month, hoy.day) < (self.fecha_nac.month, self.fecha_nac.day))
    
    def mostrar(self):
        print(f"{self.nombre} {self.apellido} - CI: {self.ci} - Edad: {self.edad()}")

class Estudiante(Persona):
    def __init__(self, ci, nombre, apellido, fecha_nac, ru, semestre):
        super().__init__(ci, nombre, apellido, fecha_nac)
        self.ru = ru
        self.semestre = semestre
    
    def mostrar(self):
        super().mostrar()
        print(f"RU: {self.ru} - Semestre: {self.semestre}")

class Docente(Persona):
    def __init__(self, ci, nombre, apellido, fecha_nac, nit, profesion, sexo):
        super().__init__(ci, nombre, apellido, fecha_nac)
        self.nit = nit
        self.profesion = profesion
        self.sexo = sexo
    
    def mostrar(self):
        super().mostrar()
        print(f"NIT: {self.nit} - Profesión: {self.profesion} - Sexo: {self.sexo}")

estudiantes = [
    Estudiante("123", "Juan", "Perez", date(1995, 5, 10), "E001", 8),
    Estudiante("456", "Maria", "Gomez", date(2000, 8, 20), "E002", 5),
    
]

docentes = [
    Docente("111", "Pedro", "Perez", date(1980, 7, 15), "N001", "Ingeniero", "M"),
    Docente("222", "Ana", "Gomez", date(1975, 3, 22), "N002", "Licenciada", "F"),
]


print("Estudiantes >25 años:")
for e in estudiantes:
    if e.edad() > 25:
        e.mostrar()

ingenieros = [d for d in docentes if d.profesion == "Ingeniero" and d.sexo == "M"]
if ingenieros:
    mayor = max(ingenieros, key=lambda x: x.edad())
    print("\nDocente Ingeniero masculino más mayor:")
    mayor.mostrar()

print("\nPersonas con mismos apellidos:")
apellidos = set(e.apellido for e in estudiantes) & set(d.apellido for d in docentes)
for a in apellidos:
    print(f"\nApellido: {a}")
    print("Estudiantes:")
    for e in [x for x in estudiantes if x.apellido == a]:
        e.mostrar()
    print("Docentes:")
    for d in [x for x in docentes if x.apellido == a]:
        d.mostrar()

File: Practica de Programacion/Practica 3 Herencia/Ejercicio 3/3 persona.java
import java.time.LocalDate;
import java.util.*;

class Persona {
    String ci, nombre, apellido;
    LocalDate fechaNac;
    
    Persona(String ci, String n, String a, LocalDate fn) {
        this.ci = ci; nombre = n; apellido = a; fechaNac = fn;
    }
    
    int edad() {
        return LocalDate.now().getYear() - fechaNac.getYear();
    }
}

class Estudiante extends Persona {
    String ru;
    Estudiante(String ci, String n, String a, LocalDate fn, String ru) {
        super(ci, n, a, fn); this.ru = ru;
    }
}

class Docente extends Persona {
    String profesion; char sexo;
    Docente(String ci, String n, String a, LocalDate fn, String p, char s) {
        super(ci, n, a, fn); profesion = p; sexo = s;
    }
}

public class Main {
    public static void main(String[] args) {
        List<Estudiante> ests = Arrays.asList(
            new Estudiante("123", "Juan", "Perez", LocalDate.of(1995,5,10), "E001"),
            new Estudiante("456", "Maria", "Gomez", LocalDate.of(2000,8,20), "E002"),
            new Estudiante("789", "Carlos", "Lopez", LocalDate.of(1998,12,5), "E003")
        );
        
        List<Docente> docs = Arrays.asList(
            new Docente("111", "Pedro", "Perez", LocalDate.of(1980,7,15), "Ingeniero", 'M'),
            new Docente("222", "Ana", "Gomez", LocalDate.of(1975,3,22), "Licenciada", 'F'),
            new Docente("333", "Luis", "Martinez", LocalDate.of(1970,11,30), "Ingeniero", 'M')
        );
        
        System.out.println("Estudiantes >25:");
        ests.stream().filter(e -> e.edad() > 25).forEach(e -> 
            System.out.println(e.nombre + " " + e.apellido));
        
        System.out.println("\nDocente Ingeniero M más viejo:");
        docs.stream()
            .filter(d -> d.profesion.equals("Ingeniero") && d.sexo == 'M')
            .max(Comparator.comparing(Persona::edad))
            .ifPresent(d -> System.out.println(d.nombre + " " + d.apellido));
        
        System.out.println("\nMismos apellidos:");
        Set<String> apeEst = new HashSet<>();
        ests.forEach(e -> apeEst.add(e.apellido));
        docs.stream()
            .filter(d -> apeEst.contains(d.apellido))
            .forEach(d -> System.out.println(d.apellido));
    }
}